design/ceph-mon-pv.md

```diff
@@ -0,0 +1,139 @@
+
+Currently all of the storage for Ceph monitors (data, logs, etc..) is provided
+using HostPath volume mounts. Supporting PV storage for Ceph monitors in
+environments with dynamically provisioned network block devices (AWS, GCE,
+etc...) will allow monitors to migrate without requiring the monitor state to be
+rebuilt. This is particularly useful in cases where nodes fails or are taken
+offline for maintenance or upgrades.
+
+There are two major components to this design. First, the cluster CRD schema is
+modified to accept a PVC template for defining the storage to be used by
+monitors. And second, monitor pod management will be handled by StatefulSets
+rather than Deployments.
+
+# Monitor storage
+
+The monitor specification in the CRD is updated to include a persistent volume
+claim template that is used to obtain PVs for the monitor data store.
+
+```go
+type MonSpec struct {
+	Count                int  `json:"count"`
+	PreferredCount       int  `json:"preferredCount"`
+	AllowMultiplePerNode bool `json:"allowMultiplePerNode"`
+	VolumeClaimTemplate  *v1.PersistentVolumeClaim
+}
+```
+
+Unlike the storage requirements of OSDs (xf: StorageClassDeviceSets) it is
+reasonable to use default requirements for monitors (e.g. 5-10 GB). This means
+that a minimal claim may be specified using only the requested storage class in
+which case default resource requests will be used.
+
+```yaml
+volumeClaimTemplates:
+- spec:
+    storageClassName: gp2-ebs
+...
+volumeClaimTemplates:
+- spec:
+    storageClassName: local-storage
+```
+
+It is also reasonable for monitors to use storage in the form of a directory on
+the host's root file system. This behavior effectively provides something like
+Rook-specific dynamic HostPath provisioning and is the current default behavior
+in Rook. In order to continue to support this level of operational simplicity,
+the volume claim template itself can be left unspecified entirely to request the
+existing default behavior. Note that the storage class `local-storage` is not
+necessarily equivalent---there are a variety of implementations with the most
+popular requiring static provisioning [0, 1].
+
+[0]: https://github.com/rancher/local-path-provisioner
+[1]: https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner
+
+## Non-PV monitor storage
+
+Not all monitor storage requirements should be satisfied using PV storage. Most
+notably log and crash data. These data may continue to be stored in HostPath
+volume mounts. This will continue to be based on `dataDirHostPath`.
+
+## Host networking
+
+Currently Ceph monitor pods may be deployed using host-based networking, or
+placed behind a service using a portable network address. This poses a challenge
+when migrating a monitor (and it's PV) to a new node which has been deployed
+using host-based networking as the monitor's address will no longer valid.
+
+Support for migration of monitors with host-based networking is future work. To
```

Comment:
Host networking is a requirement for many users so I'm not seeing that we can have this feature without it. While it feels like a step backwards from the purity of the PVC approach, don't we just need to do node selection and add a node selector similar to today?

Reply:
For host networking we can still require non-PVC storage though.

Reply:
The issue with allowing non-local PVs and host networking to operate together is that when migration occurs the ip address of the monitor binding needs to be updated (in how many places?). I'm happy to add this into the scope of the design. I think at one point we had decided against it for no other reason than to simplify the initial development.

Reply:
I was thinking that if they want host networking we would effectively disable migration by adding a node selector to the mon, similar to today. Updating the mon ip isn't something that I'd want to touch.

Reply:
Ok I had to re-read what I wrote and this conversation multiple times to figure out why weren't in agreement.

> Support for migration of monitors with host-based networking is future work.

True. And this is accomplished by what you said: using a node selector to disable migration.

But I think this discussion happened before the confusing _next_ sentence

> host-based networking will not be supported for non-local PVs.

Which is a mistake on my part.